spatial_cell_gpu.hpp

```diff
@@ -0,0 +1,1374 @@
+/*
+ * This file is part of Vlasiator.
+ * Copyright 2010-2024 Finnish Meteorological Institute and University of Helsinki
+ *
+ * For details of usage, see the COPYING file and read the "Rules of the Road"
+ * at http://www.physics.helsinki.fi/vlasiator/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+/*!
+Spatial cell class for Vlasiator that supports a variable number of velocity blocks.
+*/
+
+#ifndef VLASIATOR_SPATIAL_CELL_GPU_HPP
+#define VLASIATOR_SPATIAL_CELL_GPU_HPP
+
+#include <algorithm>
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <mpi.h>
+#include <limits>
+#include <stdint.h>
+#include <vector>
+#include <array>
+#include <unordered_map>
+#include <set>
+#include <map>
+#include <phiprof.hpp>
+#include <tuple>
+
+#include "memoryallocation.h"
+#include "common.h"
+#include "parameters.h"
+#include "definitions.h"
+
+#include "velocity_mesh_gpu.h"
+
+#include "velocity_block_container.h"
+
+#ifdef DEBUG_VLASIATOR
+   #ifndef DEBUG_SPATIAL_CELL
+   #define DEBUG_SPATIAL_CELL
+   #endif
+#endif
+
+typedef Parameters P; // Heeded in numerous files which include this one
+
+/*!
+Used as an error from functions returning velocity cells or
+as a cell that would be outside of the velocity block
+*/
+#define error_velocity_cell 0xFFFFFFFFu
+
+/*!
+Used as an error from functions returning velocity cell indices or
+as an index that would be outside of the velocity block
+*/
+#define error_velocity_cell_index 0xFFFFFFFFu
+
+namespace spatial_cell {
+
+   namespace Transfer {
```

Comment:
Are these even differing for the CPU and GPU cases? And while we're at it, shouldn't they be const static and/or an enum?

Reply:
No, those are indeed identical. 
I could foresee that with GPU-MPI-Optimizations that *might* change, but is unlikely. 
Enums give just sequential numbers, wheres here we construct a bitmask. I doubt there's a bitmask-enum-creator system?

Can make them static, though. And can place them in spatial_cell_wrapper.hpp.

Reply:
I suggest a next PR for once. :D
I tagged a number of comms patterns we don't use and won't use again ever. So these could be cleaned up. If we have an enum, we can loop over it and bitshift by that value, for example?

Reply:
Enum members [can each individually have a value assigned](https://en.cppreference.com/w/cpp/language/enum), as in:
```c++
enum foo {
  bar = 1,
  baz = 42,
  gnorf = 1337
} ;
```

Reply:
Changing the actual way we use these will be a separate thing (so can be made into an issue), but I think I'll do the static / consolidation into wrapper.hpp now.